pandas/core/generic.py

```diff
@@ -3972,6 +3972,16 @@ def asof(self, where, subset=None):
             where = Index(where) if is_list else Index([where])
 
         nulls = self.isnull() if is_series else self[subset].isnull().any(1)
+        if nulls.all():
+            if is_series:
+                return pd.Series(np.nan, index=where)
+            else:
+                if is_list:
+                    return pd.DataFrame(np.nan, index=where,
+                                        columns=self.columns)
+                else:
+                    return pd.Series(np.nan, index=[where])
```

Comment:
That is because you wrote your test without it, then of course it will pass it without it. 
The name is essential for a correct result, and should be added both to the code here as to the test.

As I showed before with this simple example of the current behaviour

```
In [1]: df = pd.DataFrame(np.random.randn(2,2), index=[1,2], columns=['A', 'B'])

In [2]: df
Out[2]: 
          A         B
1  0.387517 -0.571258
2 -0.376436  0.604668

In [3]: df.asof(3)
Out[3]: 
A   -0.376436
B    0.604668
Name: 3, dtype: float64
```

you can clearly see that the name is the actual value you passed to `asof`